src/m3/functions.ts

```diff
@@ -40,35 +53,53 @@ export const relationsOf = (metamodelElement: MetamodelElement): Link[] =>
         : []
 
 
-export type Comparer<T> = (l: T, r: T) => number
+export const containeds = (thing: M3Concept): M3Concept[] => {
+    if (thing instanceof Metamodel) {
+        return thing.elements
+    }
+    if (thing instanceof FeaturesContainer) {
+        return thing.features
+    }
+    if (thing instanceof Enumeration) {
+        return thing.literals
+    }
+    return []
+}
+
 
-const stringCompare: Comparer<string> = (l, r): number =>
-    l === r ? 0 : (l > r ? 1 : -1)
+/**
+ * Performs a depth-first tree traversal of a metamodel, "flatMapping" the `map` function on every node.
+ * It avoids visiting nodes twice (to avoid potential infinite loops), but doesn't report cycles.
+ */
+export const flatMap = <T>(metamodel: Metamodel, map: (t: M3Concept) => T[]): T[] =>
+    flatmapNonCyclingFollowing(map, containeds)(metamodel)
 
-export const stringyCompare = <T>(keyFunc: (t: T) => string): Comparer<T> =>
-    (l: T, r: T) => stringCompare(keyFunc(l), keyFunc(r))
 
-export const sortByStringKey = <T>(ts: T[], keyFunc: (t: T) => string) =>
-    [...ts].sort(stringyCompare(keyFunc))
+export const elementsSortedByName = (metamodelElements: MetamodelElement[]) =>
+    sortByStringKey(metamodelElements, (element) => element.simpleName)
 
 
-export const flatMap = <T>(metamodel: Metamodel, map: (t: M3Concept) => T[]): T[] => {
-    // (non-fancy, slightly non-FP-ish implementation of a depth-first tree traversal of a LIonCore instance)
-    const ts: T[] = []
-    const visit = (thing: M3Concept) => {
-        ts.push(...map(thing))
-        // recurse into all containments:
-        if (thing instanceof Metamodel) {
-            thing.elements.forEach(visit)
-        }
-        if (thing instanceof FeaturesContainer) {
-            thing.features.forEach(visit)
-        }
-        if (thing instanceof Enumeration) {
-            thing.literals.forEach(visit)
-        }
+export type ConceptType = Concept | ConceptInterface
```

Comment:
Does this indicate we have a missing abstraction?

Reply:
That could be, but OTOH: sum types are so cheap in TypeScript that I'm fine with not having the abstraction explicitly in the M3 until it's clear that it's missing in other/all contexts as well.